### PR TITLE
Debugger Expression Language: Remove Predicate

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do

# Motivation
Simplify expression language by removing Predicate interface and use directly Expression for Predicate instead.

# Additional Notes

### CODE REVIEW

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/IfElseExpression.java [hunk 1]
```diff
@@ -18,7 +18,7 @@ public IfElseExpression(
 
   @Override
   public Void evaluate(ValueReferenceResolver valueRefResolver) {
-    if (test.evaluate(valueRefResolver).test()) {
+    if (test.evaluate(valueRefResolver)) {
       thenExpression.evaluate(valueRefResolver);
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/IfElseExpression.java:22]: Not related to the PR - why we are not return the value from the then expression?
return the value we can do things like headers[if(response.status == 200, "Content", "Error")] like the "?:" operator.

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/ComparisonOperator.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      return Boolean.FALSE;
+    }
+  },
+  LT("<") {
+    @Override
+    public Boolean apply(Value<?> left, Value<?> right) {
+      if (left instanceof NumericValue && right instanceof NumericValue) {
+        return compare(left, right) < 0;
+      }
+      return Boolean.FALSE;
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/ComparisonOperator.java:49]: don't we support "hello" < "world" operators?

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/IfElseExpression.java (hunk 1 above)
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/IfElseExpression.java:22]: Note related - but if we return the value from thenExpression and elseExpression we can create "?:" operators.

So we can do `content[response.status == 200 ? "data" : "error"]` - not sure this is a good idea.

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/PredicateExpression.java [hunk 3]
```diff
@@ -1,10 +1,32 @@
...
 import com.datadog.debugger.el.Expression;
-import com.datadog.debugger.el.Predicate;
+import datadog.trace.bootstrap.debugger.el.ValueReferenceResolver;
 
-/** A generic interface for expressions resolving to {@linkplain Predicate} */
-public interface PredicateExpression extends Expression<Predicate> {
-  PredicateExpression TRUE = ctx -> Predicate.TRUE;
-  PredicateExpression FALSE = ctx -> Predicate.FALSE;
+/** A generic interface for expressions resolving to {@linkplain Boolean} */
+public interface PredicateExpression extends Expression<Boolean> {
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/PredicateExpression.java:7]: BooleanExpression?

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java [hunk 4]
```diff
@@ -1,89 +1,122 @@
...
-  @Test
-  void evaluateFirstUndefined() {
-    ComparisonExpression expression =
-        new ComparisonExpression(ValueExpression.UNDEFINED, new NumericValue(2), combiner);
-    assertFalse(expression.evaluate(NoopResolver.INSTANCE).test());
-    Mockito.verify(combiner, VerificationModeFactory.noInteractions())
-        .get(ArgumentMatchers.any(), ArgumentMatchers.any());
+  private static Stream<Arguments> expressions() {
+    return Stream.of(
+        Arguments.of(new NumericValue(1), new NumericValue(1), EQ, true),
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java:36]: maybe worth checking different types: NumberValue(1), NumberValue(1L), NumberValue(1.0F), NumberValue(1.0) - I think they should all be equal...

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java [hunk 5]
```diff
@@ -1,89 +1,122 @@
...
+        Arguments.of(new NumericValue(2), new NumericValue(1), LT, false),
+        Arguments.of(new NumericValue(1.0), new NumericValue(1.1), LT, true),
+        Arguments.of(ValueExpression.NULL, new NumericValue(2), LT, false),
+        Arguments.of(new NumericValue(Double.NaN), new NumericValue(Double.NaN), LT, false),
+        Arguments.of(
+            new NumericValue(BigDecimal.valueOf(1)),
+            new NumericValue(BigDecimal.valueOf(2)),
+            LT,
+            true),
+        Arguments.of(new NumericValue(1), new NumericValue(1), LE, true),
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java:74]: do we also should compare different types of numbers?

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java [hunk 6]
```diff
@@ -1,89 +1,122 @@
...
-        new ComparisonExpression(ValueExpression.UNDEFINED, new NumericValue(2), combiner);
-    assertFalse(expression.evaluate(NoopResolver.INSTANCE).test());
-    Mockito.verify(combiner, VerificationModeFactory.noInteractions())
-        .get(ArgumentMatchers.any(), ArgumentMatchers.any());
+  private static Stream<Arguments> expressions() {
+    return Stream.of(
+        Arguments.of(new NumericValue(1), new NumericValue(1), EQ, true),
+        Arguments.of(new NumericValue(1), new NumericValue(2), EQ, false),
+        Arguments.of(new StringValue("foo"), new NumericValue(2), EQ, false),
+        Arguments.of(new StringValue("foo"), new StringValue("foo"), EQ, true),
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java:39]: NumericValue(1) == StringValue("1") ?

**Review by shatzi — APPROVED** [review]
LGTM

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/IfElseExpression.java (hunk 1 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/IfElseExpression.java:22]: `then` expression is not implemented yet see https://github.com/DataDog/dd-trace-java/blob/master/dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/ThenExpression.java

File: dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/PredicateExpression.java (hunk 3 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/main/java/com/datadog/debugger/el/expressions/PredicateExpression.java:7]: done

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java (hunk 4 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java:36]: done

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java (hunk 6 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java:39]: done

File: dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java (hunk 5 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-el/src/test/java/com/datadog/debugger/el/expressions/ComparisonExpressionTest.java:78]: Not sure what will be the result and how to specify them with conversion widening, etc... rabbit holes

**Pull request merged**